SAM BOSSA bootloader support and CDC-ACM line rate change detection

nzmichaelh opened the pull request:
The BOSSA bootloader is used on Arduino compatible SAM boards like the Arduino MKR Zero, Adafruit Trinket M0, and Seeedstudio XIAO.  These boards use the 'Arduino hack' of setting the USB CDC baud rate to 1200 baud to signal the application to restart into the bootloader.

Add a new callback to the CDC-ACM UART API so the application can listen for baud rate changes.  Add a SAM0 specific module that listens for these changes and restarts into the bootloader if found.

With these changes, a "west flash" will reset the board and flash the new version.

nzmichaelh committed:
soc: sam0: add support for the BOSSA bootloader

This adds a new config option for SAM0 targets that use the BOSSA
bootloader.  If the CDC ACM driver is also enabled, then the
programmer can automatically reset the board into the bootloader for
programming.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh committed:
boards: arm: configure and document the bootloader on the Trinket M0

The Trinket M0 has a Adafruit UF2 bootloader.  Configure and document
how to use it.

Signed-off-by: Michael Hope <mlhx@google.com>

jfischer-phytec-iot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -919,6 +929,22 @@ static int cdc_acm_line_ctrl_get(struct device *dev,
+	struct cdc_acm_dev_data_t * const dev_data = DEV_DATA(dev);
+
+	if (dev_data->line_ctrl_cb != NULL) {
+		return -ENOMEM;
```
-EINVAL or -EFAULT?

nzmichaelh replied:
EINVAL SGTM, done.

jfischer-phytec-iot commented on boards/arm/adafruit_trinket_m0/doc/index.rst:
```diff
@@ -98,7 +98,12 @@ Programming and Debugging
 The Trinket M0 ships the BOSSA compatible UF2 bootloader.  The
-bootloader can be entered by quickly tapping the reset button twice.
+bootloader can be entered manually by quickly tapping the reset button
+twice, or optionally via a command over USB.
```
This somehow implies that USB device stack and CDC ACM is always present (even if one would read the text below).

nzmichaelh replied:
Reworded.  PTAL.

pabigot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -919,6 +929,22 @@ static int cdc_acm_line_ctrl_get(struct device *dev,
+	}
+
+	dev_data->line_ctrl_cb = callback;
+	dev_data->line_ctrl_cb_data = user_data;
```
This looks a little odd to me: you can pass a null callback, but only if no callback has been set.  Once a callback is set, you can't clear it.

If the intent that this function can be invoked only once, then that should be documented.  If it's possible to set the callback multiple times then a lock must be provided to ensure that use of the callback is consistent with a configured pair of callback and userdata.

Without knowing much about this I think the simplest solution is document that the function can be successfully invoked only once, and that passing a null callback returns `-EINVAL` while passing a non-null callback if one has already been set returns `-EBUSY` or `-EALREADY`.  `user_data` can be null or not, but you can't change it once the callback has been set.

If that doesn't satisfy the use case document and implement something that does.

nzmichaelh replied:
Hi Peter.  I had a look at the other callback implementations and you're right - any callback set will overwrite any previous callback set.

Changed this to match.

pabigot reviewed (dismissed):
I can't speak to the BOSSA stuff, but the API extension needs some clarification.

carlescufi commented:
@pabigot please revisit.

pabigot reviewed (commented):
The new callback still isn't managed in a thread-safe way, but neither is the existing write callback.  I don't think fixing all that is in scope for this.  So I'll remove my block, but I won't approve it at this time.  Assuming a UART/USB maintainer thinks this is OK I'll willing to second it.

nzmichaelh commented:
Hi @jfischer-phytec-iot I see you changed the title to be UART API specific.  Should I split out the BOSSA changes then?

nzmichaelh commented:
Hi @jfischer-phytec-iot @finikorg I see you're the USB subsystem owners.  What do you think of the API change?  If required, I'm happy to do a follow up PR that changes the CDC ACM implementation to register callbacks in a thread safe way.  See @pabigot 's comment for more.

jfischer-phytec-iot commented:
> Hi @jfischer-phytec-iot I see you changed the title to be UART API specific. Should I split out the BOSSA changes then?

I think not or not yet, my intention was to get more attention.

>What do you think of the API change?

+1, but I think we should discuss it in the API meeting. @carlescufi

anangl commented on include/drivers/uart.h:
```diff
@@ -1127,6 +1139,32 @@ static inline int z_impl_uart_line_ctrl_get(struct device *dev,
+	if (api && api->line_ctrl_callback_set) {
+		api->line_ctrl_callback_set(dev, callback, user_data);
+	}
+}
```
Since this API function is optional, it would be good to return `-ENOTSUP` when a given driver does not implement it.

nzmichaelh replied:
Superseded

pabigot reviewed (commented):
This seems a little under-defined.  AFAICT all you get as a user is a callback suggesting that something changed, and the user has to guess what caused the callback.

At this time the only use case is for a baud rate change.  Is there a reason to invoke it on DCD change, or CTS, or RI, or other signals? If so, does it make sense to pass to the callback the bits identifying what changed?  Then the API should document what line control state change detections are required when this feature is implemented in a driver, and which are optional.

If not, should this just be a baud rate changed callback, and pass the new baud rate as a parameter?

pabigot commented:
Notes from 2020-06-30 API telecon.  Edit if I missed something.

JF: Arduino BOSSA uses changes of baud rate to determine whether system should trigger boot loader.  Do we need this?

KC: Seems limited

JF: Likely that other signals than baud rate could be used.

PB: Not sure hardware would support notification of change in baud rate even if autobaud supported.  Similar for low-level control signals.

Summary: Need more input to determine whether this is useful, what would be useful, and whether any of it is implementable on anything other than CDC-ACM.  If not this may be suitable as a device-specific extension for CDC-ACM somehow.

nzmichaelh commented:
Hi Peter.  There's two PRs for adding BOSSA / Arduino reset support so far - this and #25422

#25422 directly modifies cdc_acm and the feedback was that it was too tightly coupled and should be a generic API.  This PR is a generic API, but the feedback is that there's not enough use cases to add it.

We're getting conflicting feedback.  How should we proceed?

jfischer-phytec-iot commented:
> If not this may be suitable as a device-specific extension for CDC-ACM somehow.

> We're getting conflicting feedback. How should we proceed?

I guess something was misunderstood. From my side, only this proposal makes sense, not the hack in #25422

pabigot commented:
> We're getting conflicting feedback. How should we proceed?

Sorry; like many things in Zephyr what at first seems simple gets complicated.

My position is that if we're adding new API to the generic UART driver we need to know the requirements for the API.  `uart_line_ctrl_callback_set()` needs justification, more clear definition, and confirmation that it can be implemented on multiple platforms.

I understand that the goal is:

> Detect an Arduino-style upgrade request and implement the required response

The Zephyr implementation should respect these distinct phases.  I agree with @jfischer-phytec-iot that the approach in #25422 is not acceptable: it only works with one platform on one driver, and puts target-specific upgrade logic in a serial driver.  If somebody has a non-SAM0 Arduino-style board it won't help them, and maintaining the serial driver becomes complex.

So my first question is: Can any UART driver other than CDC-ACM detect this event?  If so, what other line-control events can be detected, and how must the subsystem `uart_something_changed()` API be defined to support them?

In answering that I've now learned about [BOSSA](http://www.shumatech.com/web/products/bossa) which is a red-herring, and  about how original Arduino hardware [tied DTR to the MCU RESETn signal](https://liudr.wordpress.com/2016/02/26/how-arduino-bootloader-works/) to force a reboot, and then how for USB where DTR is not a physical signal this is implemented by [opening the serial port at 1200 baud and then closing it](https://arduino.stackexchange.com/a/32737).

So it appears that this functionality is only ever useful for CDC_ACM UARTs, and for wired RS-232 there's no software needed because the reset is done in hardware.

So: IMO the line state change detection API doesn't belong in the UART API.

In that case, I would propose to add an an option `CONFIG_CDC_ACM_ENABLE_ARDUINO_RESET` flag, which puts the detection logic into the cdc_acm driver and defines a non-syscall `cdc_acm_arduino_reset_callback(dev, cb)` function that installs a callback to be used when the event occurs.  You'll need to add a small header in `install/drivers/uart/cdc_acm.h` to provide the definition.

Reading the links above I'm not convinced the response to the event needs to erase the first word of the application: it seems it's just supposed to be a soft reset, which starts the bootloader, which may or may not perform an update.  I may be wrong in that interpretation.

But in any case the most flexible approach is just detecting the situation with API like the one two paragraphs up, and letting the application determine what "reset" means.  (If necessary I suppose a `lib/arduino` could be added with `arduino_reset_to_upgrade()` implemented for shared use, but that should be a separate PR from adding the ability to detect the request.)

@jfischer-phytec-iot Does this plan make sense to you?

nzmichaelh commented:
> Sorry; like many things in Zephyr what at first seems simple gets complicated.

It's all good.  Thanks for the detailed reply.

> My position is that if we're adding new API to the generic UART driver we need to know the requirements for the API. `uart_line_ctrl_callback_set()` needs justification, more clear definition, and confirmation that it can be implemented on multiple platforms.
> 
> I understand that the goal is:
> 
> > Detect an Arduino-style upgrade request and implement the required response
> 
> The Zephyr implementation should respect these distinct phases. I agree with @jfischer-phytec-iot that the approach in #25422 is not acceptable: it only works with one platform on one driver, and puts target-specific upgrade logic in a serial driver. If somebody has a non-SAM0 Arduino-style board it won't help them, and maintaining the serial driver becomes complex.
> 
> So my first question is: Can any UART driver other than CDC-ACM detect this event? If so, what other line-control events can be detected, and how must the subsystem `uart_something_changed()` API be defined to support them?

Unfortunately I don't have other concrete use cases.  Other serial emulation protocols such as Bluetooth SPP or RFC2217 have similar features where the line control and configuration is set remotely.

I can imagine using this for a Zephyr based USB -> Serial adapter, but that's a hypothetical.
 
> In answering that I've now learned about [BOSSA](http://www.shumatech.com/web/products/bossa) which is a red-herring, and about how original Arduino hardware [tied DTR to the MCU RESETn signal](https://liudr.wordpress.com/2016/02/26/how-arduino-bootloader-works/) to force a reboot, and then how for USB where DTR is not a physical signal this is implemented by [opening the serial port at 1200 baud and then closing it](https://arduino.stackexchange.com/a/32737).
> 
> So it appears that this functionality is only ever useful for CDC_ACM UARTs, and for wired RS-232 there's no software needed because the reset is done in hardware.

Correct.  We can ignore the RS-232 case as well as no modern boards use it.
 
> So: IMO the line state change detection API doesn't belong in the UART API.
> 
> In that case, I would propose to add an an option `CONFIG_CDC_ACM_ENABLE_ARDUINO_RESET` flag, which puts the detection logic into the cdc_acm driver and defines a non-syscall `cdc_acm_arduino_reset_callback(dev, cb)` function that installs a callback to be used when the event occurs. You'll need to add a small header in `include/drivers/uart/cdc_acm.h` to provide the ~definition~ declaration.

SGTM.

BTW, another alternative is adding a `SYS_REBOOT_INTO_BOOTLOADER` in `power/reboot.h`, but then cdc_acm has specific knowledge about the Arduino bootloader.
 
> Reading the links above I'm not convinced the response to the event needs to erase the first word of the application: it seems it's just supposed to be a soft reset, which starts the bootloader, which may or may not perform an update. I may be wrong in that interpretation.

I agree.  Recent bootloaders will stay in the bootloader if there's a magic value at the top of RAM, and this PR uses that method.
 
> But in any case the most flexible approach is just detecting the situation with API like the one two paragraphs up, and letting the application determine what "reset" means. (If necessary I suppose a `lib/arduino` could be added with `arduino_reset_to_upgrade()` implemented for shared use, but that should be a separate PR from adding the ability to detect the request.)

The reset code is very SOC specific and also tiny, so I'd duplicate it for other non-SAM0 SOCs if it's ever needed.
 
> @jfischer-phytec-iot Does this plan make sense to you?

nzmichaelh committed:
usb: cdc_acm: add a DTE rate changed callback

This is called when the USB host changes the virtual baud rate.

This is used by Arduino style boards to signal that the board should
reset into the bootloader.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
Changed to a CDC ACM specific callback.  PTAL.

pabigot commented on include/usb/class/usb_cdc.h:
```diff
@@ -165,4 +165,26 @@ struct cdc_ecm_descriptor {
+ *
+ * @param port Device struct for the CDC ACM device.
+ */
+typedef void (*cdc_acm_arduino_reset_callback_t)(struct device *port);
```
But "arduino reset" is an event that, based on [my analysis](https://github.com/zephyrproject-rtos/zephyr/pull/26173#issuecomment-652353731)  can only be detected in the CDC ACM driver.

> It is similar attempt as #25422 and I will not agree with any attempt to add a callback in the control request processing.

Do you have an alternative method to detect the event of a baud rate change to 1200?

jfischer-phytec-iot replied:
Not relevant, this is a header for CDC (ACM) class specification, whatever arduino-callback is not part of it.

pabigot replied:
There I agree with @jfischer-phytec-iot.  I asked that this be put in `include/drivers/uart/cdc_acm.h` which is the correct path for driver-specific extensions to the CDC-ACM UART driver.

If this moves, are you more willing to accept this driver-specific extension?

nzmichaelh replied:
Moved

pabigot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -228,11 +232,17 @@ int cdc_acm_class_handle_req(struct usb_setup_packet *pSetup,
 			dev_data->line_coding.bParityType,
 			dev_data->line_coding.bDataBits);
+#if defined(CONFIG_CDC_ACM_ARDUINO_RESET_SUPPORT)
+		if (rate == 1200 && dev_data->reset_cb != NULL) {
```
I should be clear: I'm fine with generalizing this from detecting a change to 1200 baud to detecting a change in baud rate from the last observed (or configured) value.  But unless there's evidence this can be done on other UART platforms I believe this should be a CDC-ACM--specific extension, and I don't see any reason to generalize from baud rate change to arbitrary line control change.

nzmichaelh replied:
As an alternative, I've changed this to `cdc_acm_dte_rate_callback_set` that is called when the USB host changes the baud rate.
The only use case is resetting into the bootloader, but now the CDC ACM code has nothing Arduino specific in it.

nzmichaelh replied:
PTAL

carlescufi replied:
I actually agree with the proposed change by @nzmichaelh here. I know that for example, the J-Link CDC ACM implementation inside our debugger chips use the exact same logic (baudrate change from Host) to detect the fact that the UART has been "connected to" from the Host, i.e. that a program has opened it.

@jfischer-phytec-iot what is the argument here against this current approach?

carlescufi replied:
>It is pure generic code, and then there is this very specific to a platform(IDE?) and architecture. IMHO, we should generally not allow such a thing. Which specific IDE bootloader method is next?

No, this is actually the only reliable way a USB CDC ACM implementation can detect that a Host has opened a serial port, as far as I know.

pabigot replied:
Note that the use case and implementation here does not detect that a host has opened a serial port; it detects that the host has opened a serial port *at a different baud rate*.  I suppose this could be converted to a "serial-port-opened" callback, if there's a reason to provide that.

I do have a use case for that sort of thing but was intending to handle it based on notifications of whether USB was connected or not, though I haven't had time to look into how that can be made to work.

jfischer-phytec-iot replied:
> I do have a use case for that sort of thing but was intending to handle it based on notifications of whether USB was connected or not, though I haven't had time to look into how that can be made to work.

You can use DTR for that, see samples/subsys/usb/cdc_acm

jfischer-phytec-iot replied:
> I actually agree with the proposed change by @nzmichaelh here. I know that for example, the J-Link CDC ACM implementation inside our debugger chips use the exact same logic (baudrate change from Host) to detect the fact that the UART has been "connected to" from the Host, i.e. that a program has opened it.

> 
> @jfischer-phytec-iot what is the argument here against this current approach?

Generally I find this approach to go into bootloader mode badly designed. It is not that there are no other methods, for example USB DFU. Here whatever code is executed in the context of USB controll request, for me it is not right, but I probably would have no problem to use simple IPC (like DBUS) for that. This code is inserted and needs to be maintained, just for the reason that some IDE can trigger the bootloader for some types of boards without pressing the button. By the way, it only works if USB is included in the firmware.
That is why I do not like this approach. I have left some comments, otherwise I leave the decision if it should go in or not to others.

nzmichaelh replied:
Thanks Johann.  To clarify - this isn't for a new bootloader design, but for compatibility with the manufacturer's default bootloader.  I agree if we were starting from a greenfield then something like DFU is better.

carlescufi replied:
>You can use DTR for that, see samples/subsys/usb/cdc_acm

I am not sure that works on all platforms @jfischer-phytec-iot. From experience with J-Link's CDC ACM, I have found that only a baud rate **change** is reliable in getting through on all OSs. CC @emob-nordic. Maybe that's why this BOSSA bootloader uses this approach.

Unfortunately we don't have a system-wide notification message today so that users can subscribe and listen to baud rate (or any other hardware parameters) notifications. In the meantime, I would argue that this has the potential to be useful for others and I'd vote for its inclusion.

jfischer-phytec-iot replied:
DTR (and RTS) over SetControlLineState is what the spec offers, SetLineCoding is misuse.

carlescufi replied:
> DTR (and RTS) over SetControlLineState is what the spec offers, SetLineCoding is misuse.

I agree it's a misuse, but it's a practical misuse, a quirk.

jfischer-phytec-iot replied:
Yeah, like running a red light.

pabigot commented on soc/arm/atmel_sam0/common/bossa.c:
```diff
@@ -0,0 +1,57 @@
+static int bossa_init(struct device *unused)
+{
+	struct device *dev =
+		device_get_binding(CONFIG_USB_CDC_ACM_DEVICE_NAME "_0");
```
How do you know this is the right device to use?  Possibly replace by `CONFIG_UART_CONSOLE_ON_DEV_NAME` as there doesn't appear to be any coupling between devicetree identifiers and USB devices.

Then check whether the reset callback succeeds.

nzmichaelh replied:
The console alias is no good, as it's common to have the console on a physical UART.

Leaving open while I think about it.

pabigot replied:
How about a Kconfig `BOSSA_UART_LABEL`, defaulted to `"CDC_ACM_0"`?   (That the label is always for a CDC-ACM UART doesn't change the fact that from the perspective of the protocol it's a UART, though the limitation could be noted in the help section of the option.)

nzmichaelh replied:
I went for `_DEVICE_NAME` to match the style of `CONFIG_USB_CDC_DEVICE_NAME`

pabigot commented on include/usb/class/usb_cdc.h:
```diff
@@ -165,4 +165,26 @@ struct cdc_ecm_descriptor {
+ * @param dev       CDC ACM device structure.
+ * @param callback  Event handler.
+ */
+void cdc_acm_arduino_reset_callback_set(
```
This needs an `int` return value that can be provide `-ENOSYS`  (or `-ENOTSUP`) if something goes wrong.

nzmichaelh replied:
Done.

Note that I modeled this on the other callback setters which also return void.

pabigot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -745,6 +755,16 @@ static void cdc_acm_irq_callback_set(struct device *dev,
+void cdc_acm_arduino_reset_callback_set(
+	struct device *dev, cdc_acm_arduino_reset_callback_t callback)
+{
+	struct cdc_acm_dev_data_t *const dev_data = DEV_DATA(dev);
```
Verify that `dev->driver_api` is equal to `cdc_acm_driver_api` and return an error if it isn't.  Otherwise passing the wrong device will corrupt something else's state.

nzmichaelh replied:
Done.

pabigot reviewed (changes requested):
I'm OK with this if the CDC-ACM API extensions are moved to `include/drivers/uart/cdc_acm.h` and the driver passed to the extension is checked for validity.

pabigot reviewed (changes requested):
OK; I like this better as it's more general.  One fix for byte order requested.

tbursztyka commented:
@nzmichaelh Can you edit your PR title?

nzmichaelh commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -226,13 +233,20 @@ int cdc_acm_class_handle_req(struct usb_setup_packet *pSetup,
 
 	switch (pSetup->bRequest) {
 	case SET_LINE_CODING:
+		rate = dev_data->line_coding.dwDTERate;
```
Good catch.  I haven't tested this yet.  My plan is to wait on feedback on the structure before going further.

Leaving open until tested.

nzmichaelh replied:
Done and tested on the Trinket M0

pabigot commented:
> @nzmichaelh Can you edit your PR title?

Good catch; the previous text was confusing given the current content.  I also updated the summary slightly to clarify that this is no longer proposing a change to the UART subsystem API, just extending CDC-ACM UART device.

jfischer-phytec-iot commented on include/drivers/uart/cdc_acm.h:
```diff
@@ -0,0 +1,41 @@
+typedef void (*cdc_acm_dte_rate_callback_t)(struct device *dev, uint32_t rate);
+
+/**
+ * @brief Set the callback for DTE rate change events.
```
s/DTE rate change event/dwDTERate SetLineCoding request
s/virtual baud/baud/g

nzmichaelh replied:
Done

jfischer-phytec-iot commented on include/drivers/uart/cdc_acm.h:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @param port Device struct for the CDC ACM device.
+ */
+typedef void (*cdc_acm_dte_rate_callback_t)(struct device *dev, uint32_t rate);
```
s/cdc_acm_dte_rate_callback_t/cdc_dte_rate_callback_t/g

nzmichaelh replied:
Done

nzmichaelh commented:
Thanks for all the comments everyone.  The remaining issue is identifying the CDC ACM device name, but otherwise any objections?

carlescufi commented:
> Thanks for all the comments everyone. The remaining issue is identifying the CDC ACM device name, but otherwise any objections?

That's correct, thanks!

pabigot commented on Kconfig.zephyr:
```diff
@@ -399,6 +399,41 @@ config BOOTLOADER_ESP_IDF
+	  ACM USB support is also enabled then the board will reboot into the
+	  bootloader automatically when bossac is run.
+
+config BOOTLOADER_BOSSA_DEVICE
```
This should be `BOOTLOADER_BOSSA_DEVICE_NAME` (or `..._DEV_NAME`, which seems a little more common.

```
git grep -B1 -i 'string.*device name'
```

Fixup will also need to be squashed into the introducing commit.

nzmichaelh replied:
Done.  Kept DEVICE_NAME to match CDC_ACM_DEVICE_NAME

pabigot commented on soc/arm/atmel_sam0/common/bossa.c:
```diff
@@ -0,0 +1,60 @@
+
+SYS_INIT(bossa_init, APPLICATION, 0);
+
+#endif /* CONFIG_BOOTLOADER_BOSSA_DEVICE */
```
FYI this not being `CONFIG_BOOTLOADER_BOSSA_DEVICE_NAME` is why compliance check failed (this is an unrecognized Kconfig name).

nzmichaelh replied:
Done

nzmichaelh commented:
Hi folks.  The CDC ACM device name issue has been addressed and [1] defers the decision on this landing to others.

What's the next step here?

[1]: https://github.com/zephyrproject-rtos/zephyr/pull/26173#discussion_r452374618

nashif commented on Kconfig.zephyr:
```diff
@@ -399,6 +399,41 @@ config BOOTLOADER_ESP_IDF
 	  inside the build folder.
 	  At flash time, the bootloader will be flashed with the zephyr image
 
+config BOOTLOADER_BOSSA
```
I feel we do need some other place for all those bootloader options, the top level Kconfig.zephyr file does not seem to be the right place for this.

You are just adding into something that already exists, so we can fix this later I guess.

carlescufi commented:
@jfischer-phytec-iot can you please approve?
